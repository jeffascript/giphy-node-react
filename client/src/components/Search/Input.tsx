import React from 'react';
import { Icon, InputGroup, Tooltip, Whisper } from 'rsuite';
import { VscClearAll } from 'react-icons/vsc';
import ActionButton from '../Button/ActionButton';
import { FlexContainer, StyledInputGroup, StyledInput } from './searchbox.styles';

interface Input {
    state: string;
    handleOnchange: (e: string) => void;
    fetchGifs: (e: any, data: string) => void;
    clearData: () => void;
}

export const Input: React.FC<Input> = ({ state, handleOnchange, fetchGifs, clearData }) => {
    return (
        <>
            <FlexContainer>
                <Whisper
                    trigger="focus"
                    speaker={<Tooltip>{state?.length < 2 ? 'Minimum of 3 letters Required' : null}</Tooltip>}
                    placement={'bottomStart'}
                >
                    <StyledInputGroup inside={true}>
                        <StyledInput
                            onChange={(e) => handleOnchange(e)}
                            value={state}
                            onPressEnter={(e) => fetchGifs(e, state)}
                            size={'lg'}
                            placeholder={'Start typing ... e.g: happy'}
                        />
                        <InputGroup.Button onClick={(e) => fetchGifs(e, state)}>
                            <Icon icon="search" />
                        </InputGroup.Button>
                    </StyledInputGroup>
                </Whisper>
                <ActionButton
                    className="d-flex text-center justify-content-center align-items-center m-auto"
                    onClick={clearData}
                    data-testid="cleardata"
                >
                    <VscClearAll size={22} /> &nbsp; <span>reset</span>
                </ActionButton>
            </FlexContainer>
        </>
    );
};
