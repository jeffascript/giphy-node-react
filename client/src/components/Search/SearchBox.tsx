import React, { useCallback, useState } from 'react';
import { Icon, InputGroup, Tooltip, Whisper } from 'rsuite';
import { VscClearAll } from 'react-icons/vsc';
import { useSearchContext } from '../../ContextAPI/SearchHookContext';

import ActionButton from '../Button/ActionButton';
import { FlexContainer, StyledInputGroup, StyledInput } from './searchbox.styles';
import { ActionType } from '../../ContextAPI/Actions.context';

const SearchBox: React.FC = React.memo(() => {
    const { context, fetchWithHooks } = useSearchContext();

    const [state, setState] = useState<string>('');
    const [resultString, setResultString] = useState<string>('');

    const fetchGifs = async () => {
        fetchWithHooks(state);
        setResultString(state);
    };

    const clearData = useCallback(() => {
        context.dispatch({ type: ActionType.Clear });
        setState('');
        setResultString('');
    }, []);

    return (
        <>
            <FlexContainer>
                <Whisper
                    trigger="focus"
                    speaker={<Tooltip>{state.length < 2 ? 'Minimum of 3 letters Required' : null}</Tooltip>}
                    placement={'bottomStart'}
                >
                    <StyledInputGroup inside={true}>
                        <StyledInput
                            onChange={(e) => setState(e)}
                            value={state}
                            onPressEnter={fetchGifs}
                            size={'lg'}
                            placeholder={'Start typing ... e.g: happy'}
                        />
                        <InputGroup.Button onClick={fetchGifs}>
                            <Icon icon="search" />
                        </InputGroup.Button>
                    </StyledInputGroup>
                </Whisper>
                <ActionButton
                    className="d-flex text-center justify-content-center align-items-center m-auto"
                    onClick={clearData}
                >
                    <VscClearAll size={22} /> &nbsp; <span>reset</span>
                </ActionButton>
            </FlexContainer>
            {resultString && (
                <div className="d-flex text-center justify-content-center align-items-center  mb-5">
                    Result for: &nbsp; <h3 style={{ color: 'var(--app-green)', fontWeight: 'bold' }}>{resultString}</h3>
                </div>
            )}
        </>
    );
});

export default SearchBox;
