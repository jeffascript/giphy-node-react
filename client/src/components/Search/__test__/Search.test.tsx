/* eslint-disable no-irregular-whitespace */
import React from 'react';
import ReactDOM from 'react-dom';

import { render, screen, fireEvent, cleanup, getByTestId, queryByTestId, waitFor } from '@testing-library/react';

import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SearchBox from '../SearchBox';
import { Input } from '../Input';

Enzyme.configure({ adapter: new Adapter() });

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SearchBox />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('SearchBar Component loads without any string on the Result String', () => {
    const wrapper = shallow(<SearchBox />);
    const text = wrapper.find({ prop: 'data-testid' }).first();
    expect(text).toMatchObject({});
});

afterEach(cleanup);

it('SearchBar Component loads without any string on the Result String', () => {
    const { container } = render(<SearchBox />);
    const input = screen.getByRole('textbox');
    expect(input.textContent).toEqual('');
});

const changeHandler = jest.fn();
const clearDataHandler = jest.fn();
const fetchGifsHandler = jest.fn();

const state = 'hello';

test('Data input can be cleared', () => {
    const { rerender, asFragment, container, getByText } = render(
        <Input
            state={state}
            fetchGifs={fetchGifsHandler}
            clearData={clearDataHandler}
            handleOnchange={changeHandler}
        />,
    );

    const clearBtn = screen.getByText('reset');

    const textInput = screen.getByRole('textbox');
    expect(textInput).toHaveValue(state);
    fireEvent.change(textInput, { target: { value: 'world' } });

    fireEvent.click(clearBtn);

    expect(clearDataHandler).toHaveBeenCalled();
    rerender(
        <Input state={''} fetchGifs={fetchGifsHandler} clearData={clearDataHandler} handleOnchange={changeHandler} />,
    );
    expect(textInput).toHaveValue('');
});

it('Matches Inline snapshot', () => {
    const { rerender, asFragment, container, getByText } = render(<SearchBox />);
    expect(asFragment()).toMatchInlineSnapshot(`
        <DocumentFragment>
          <div
            class="css-3l4gaw"
          >
            <div
              class="rs-input-group css-1fesyhj rs-input-group-inside"
            >
              <input
                class="rs-input css-1rxzna3 rs-input-lg"
                placeholder="Start typing ... e.g: happy"
                type="text"
                value=""
              />
              <a
                class="rs-btn rs-btn-default rs-input-group-btn"
              >
                <i
                  class="rs-icon rs-icon-search"
                />
                <span
                  class="rs-ripple-pond"
                >
                  <span
                    class="rs-ripple"
                  />
                </span>
              </a>
            </div>
            <div
              class="d-flex text-center justify-content-center align-items-center m-auto"
              style="cursor: pointer;"
            >
              <svg
                fill="currentColor"
                height="22"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="22"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 12.6l.7.7 1.6-1.6 1.6 1.6.8-.7L13 11l1.7-1.6-.8-.8-1.6 1.7-1.6-1.7-.7.8 1.6 1.6-1.6 1.6zM1 4h14V3H1v1zm0 3h14V6H1v1zm8 2.5V9H1v1h8v-.5zM9 13v-1H1v1h8z"
                />
              </svg>
               Â  
              <span>
                reset
              </span>
            </div>
          </div>
        </DocumentFragment>
    `);
});

test('Click on Submit button and the String response reacts accordingly', async () => {
    const { rerender, asFragment, container, getByText } = render(<SearchBox />);

    expect(screen.queryByText('Result for:')).toBe(null);
    const btn = container.querySelector('a.rs-btn.rs-btn-default.rs-input-group-btn') || screen.getByRole('button');
    const textInput = screen.getByRole('textbox');
    const newValue = 'search';
    fireEvent.change(textInput, { target: { value: newValue } });
    expect(textInput).toHaveValue(newValue);
    fireEvent.submit(textInput);
    fireEvent.click(btn);
    rerender(<SearchBox />);

    expect(screen.queryByText('Result for:')).not.toBeInTheDocument();
    await waitFor(() => {
        expect(screen.queryByText('Result for:')).not.toBeInTheDocument();
    });
});

it('renders < SearchBox/> without crashing  ', () => {
    const { container } = render(<SearchBox />);
    const searchPlaceholder = 'Start typing ... e.g: happy';

    expect(screen.getAllByRole('textbox')).toHaveLength(1);
});

it('Matches snapshot', () => {
    const tree = renderer.create(<SearchBox />);
    expect(tree).toMatchSnapshot();
});
