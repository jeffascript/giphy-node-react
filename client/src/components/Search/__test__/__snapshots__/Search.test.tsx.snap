// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
Array [
  <div
    className="css-3l4gaw"
  >
    <div
      className="rs-input-group css-1fesyhj rs-input-group-inside"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <input
        className="rs-input css-1rxzna3 rs-input-lg"
        id={null}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Start typing ... e.g: happy"
        type="text"
        value=""
      />
      <a
        className="rs-btn rs-btn-default rs-input-group-btn"
        onClick={[Function]}
      >
        <i
          className="rs-icon rs-icon-search"
        />
        <span
          className="rs-ripple-pond"
        >
          <span
            className="rs-ripple"
            style={Object {}}
          />
        </span>
      </a>
    </div>
    <div
      className="d-flex text-center justify-content-center align-items-center m-auto"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <svg
        fill="currentColor"
        height={22}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width={22}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 12.6l.7.7 1.6-1.6 1.6 1.6.8-.7L13 11l1.7-1.6-.8-.8-1.6 1.7-1.6-1.7-.7.8 1.6 1.6-1.6 1.6zM1 4h14V3H1v1zm0 3h14V6H1v1zm8 2.5V9H1v1h8v-.5zM9 13v-1H1v1h8z"
        />
      </svg>
       Â  
      <span>
        reset
      </span>
    </div>
  </div>,
  "",
]
`;
